File: 01-1000/051-100/054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        # enums the old... annoying way
        TOP = 0
        RIGHT = 1
        BOTTOM = 2
        LEFT = 3
    
        current_side = TOP
        ret = []
    
        bounds = [0] * 4
        bounds[RIGHT] = len(matrix[0])
        bounds[BOTTOM] = len(matrix)
    
        while bounds[LEFT] < bounds[RIGHT] and bounds[TOP] < bounds[BOTTOM]:
            # this if elif mess should be a match case block
            # but python makes this do annyoing to do
            if current_side == TOP:
                ret += matrix[bounds[TOP]][bounds[LEFT] : bounds[RIGHT]]
                bounds[TOP] += 1
                current_side = RIGHT
    
            elif current_side == RIGHT:
                right_side = (matrix[i][bounds[RIGHT] - 1] for i in range(bounds[TOP], bounds[BOTTOM]))
                ret.extend(right_side)
                bounds[RIGHT] -= 1
                current_side = BOTTOM
    
            elif current_side == BOTTOM:
                ret += reversed(matrix[bounds[BOTTOM] - 1][bounds[LEFT] : bounds[RIGHT]])
                bounds[BOTTOM] -= 1
                current_side = LEFT
    
            elif current_side == LEFT:
                left_side = (
                    matrix[i][bounds[LEFT]]
                    for i in reversed(range(bounds[TOP], bounds[BOTTOM]))
                )
                ret.extend(left_side)
                bounds[LEFT] += 1
                current_side = TOP
        
        return ret
    

File: 01-1000/051-100/054-spiral-matrix.rs
impl Solution {
    pub fn spiral_order(matrix: Vec<Vec<i32>>) -> Vec<i32> {
        enum Side {
            Top, Bottom, Left, Right
        }

        let mut cur_side = Side::Top;

        // going to slowly circle in, reducing the bounds
        let mut top_bound = 0;
        let mut bottom_bound = matrix.len();
        let mut left_bound = 0;
        let mut right_bound = matrix[0].len();

        let mut ret = Vec::with_capacity(bottom_bound * right_bound);

        while (top_bound < bottom_bound) && (left_bound < right_bound) {
            match cur_side {
                Side::Top => {
                    ret.extend_from_slice(&matrix[top_bound][left_bound..right_bound]);
                    top_bound += 1;
                    cur_side = Side::Right;
                },
                Side::Bottom => {
                    ret.extend(matrix[bottom_bound - 1][left_bound..right_bound].iter().rev());
                    bottom_bound -= 1;
                    cur_side = Side::Left;
                },
                Side::Left => {
                    let col_iter = (top_bound..bottom_bound).rev().map(|n| matrix[n][left_bound]);
                    ret.extend(col_iter);
                    left_bound += 1;
                    cur_side = Side::Top;
                },
                Side::Right => {
                    let col_iter = (top_bound..bottom_bound).map(|n| matrix[n][right_bound - 1]);
                    ret.extend(col_iter);
                    right_bound -= 1;
                    cur_side = Side::Bottom;
                },
            }
        }

        ret
    }
}
